File: bench/src/analytics/analytics.js
"use strict";
const fs=require("fs");
const readline=require("readline");
const inputFile=process.argv[2]||"bench/data/sales.csv";
let tr=0,rev=0,units=0;
const rr=new Map(),rp=new Map(),rc=new Map();
const rl=readline.createInterface({input:fs.createReadStream(inputFile),crlfDelay:Infinity});
let header=true;
rl.on("line",(line)=>{
 if(header){header=false;return;}
 const c=line.split(",");
 if(c.length<12)return;
 tr++;
 const u=parseInt(c[7],10);
 const t=parseFloat(c[10]);
 rev+=t;
 units+=u;
 const region=c[2],country=c[3],product=c[4];
 rr.set(region,(rr.get(region)||0)+t);
 rp.set(product,(rp.get(product)||0)+t);
 rc.set(country,(rc.get(country)||0)+t);
});
rl.on("close",()=>{
 const avg=tr>0?rev/tr:0;
 const srt=(a,b)=>b[1]-a[1]||(a[0]<b[0]?-1:a[0]>b[0]?1:0);
 const rg=[...rr.entries()].sort(srt);
 const pr=[...rp.entries()].sort(srt).slice(0,5);
 const ct=[...rc.entries()].sort(srt).slice(0,3);
 const o=["=== DATA ANALYTICS RESULTS ===",`total_rows: ${tr}`,`total_revenue: ${rev.toFixed(2)}`,`avg_order: ${avg.toFixed(2)}`,`total_units: ${units}`];
 for(let i=0;i<rg.length;i++)o.push(`region_${i+1}: ${rg[i][0]} $${rg[i][1].toFixed(2)}`);
 for(let i=0;i<pr.length;i++)o.push(`product_${i+1}: ${pr[i][0]} $${pr[i][1].toFixed(2)}`);
 for(let i=0;i<ct.length;i++)o.push(`country_${i+1}: ${ct[i][0]} $${ct[i][1].toFixed(2)}`);
 process.stdout.write(o.join("\n")+"\n");
});

File: bench/src/analytics/analytics.cpp
#include<algorithm>
#include<cstdint>
#include<cstdio>
#include<cstdlib>
#include<string>
#include<unordered_map>
#include<utility>
#include<vector>
int main(int argc,char*argv[]){
 const char*path=argc>1?argv[1]:"bench/data/sales.csv";
 FILE*fp=std::fopen(path,"rb");
 if(!fp){std::fprintf(stderr,"Error: cannot open %s\n",path);return 1;}
 std::setvbuf(fp,nullptr,_IOFBF,1<<20);
 char line[8192];
 std::fgets(line,sizeof(line),fp);
 std::unordered_map<std::string,double> rr,pr,cr;
 rr.reserve(1<<16);pr.reserve(1<<16);cr.reserve(1<<16);
 uint64_t tr=0,tu=0;
 double trev=0.0;
 while(std::fgets(line,sizeof(line),fp)){
  char*p=line;
  char*fields[12]={nullptr};
  int n=0;
  fields[0]=p;
  while(*p&&n<11){if(*p==','){*p='\0';n++;fields[n]=p+1;}p++;}
  if(n<11)continue;
  const char*region=fields[2];
  const char*country=fields[3];
  const char*product=fields[4];
  uint64_t units=static_cast<uint64_t>(std::strtoll(fields[7],nullptr,10));
  double total=std::strtod(fields[10],nullptr);
  tr++;trev+=total;tu+=units;
  rr[region]+=total;
  pr[product]+=total;
  cr[country]+=total;
 }
 std::fclose(fp);
 auto cmp=[](const std::pair<std::string,double>&a,const std::pair<std::string,double>&b){
  if(a.second!=b.second)return a.second>b.second;
  return a.first<b.first;
 };
 std::vector<std::pair<std::string,double>> regions(rr.begin(),rr.end());
 std::vector<std::pair<std::string,double>> products(pr.begin(),pr.end());
 std::vector<std::pair<std::string,double>> countries(cr.begin(),cr.end());
 std::sort(regions.begin(),regions.end(),cmp);
 std::sort(products.begin(),products.end(),cmp);
 std::sort(countries.begin(),countries.end(),cmp);
 double avg=tr>0?trev/static_cast<double>(tr):0.0;
 std::printf("=== DATA ANALYTICS RESULTS ===\n");
 std::printf("total_rows: %llu\n",static_cast<unsigned long long>(tr));
 std::printf("total_revenue: %.2f\n",trev);
 std::printf("avg_order: %.2f\n",avg);
 std::printf("total_units: %llu\n",static_cast<unsigned long long>(tu));
 int nr=static_cast<int>(regions.size());
 for(int i=0;i<nr;++i)std::printf("region_%d: %s $%.2f\n",i+1,regions[i].first.c_str(),regions[i].second);
 int np=static_cast<int>(std::min<size_t>(5,products.size()));
 for(int i=0;i<np;++i)std::printf("product_%d: %s $%.2f\n",i+1,products[i].first.c_str(),products[i].second);
 int nc=static_cast<int>(std::min<size_t>(3,countries.size()));
 for(int i=0;i<nc;++i)std::printf("country_%d: %s $%.2f\n",i+1,countries[i].first.c_str(),countries[i].second);
 return 0;
}
